### DISCUSSION TITLE
Dimension text has wrong alignment in export by matplotlib backend

### DISCUSSION THREAD

**Discussion by k3rn3l3rr0r** [mozman/ezdxf#319] [Q&A] [answered]
First of all, many thanks to Manfred for creating and maintaining this great library!

I'm using the dimlines addon to create dimension annotations before generating quick PDF preview of the DXF file contents for the user via the matplotlib backend. The dimension text is properly centred in the DXF, but if exported into PDF the text is not centred to the middle of the dimension line and instead starts from the middle and continues to the right of the center.

Is there any simple solution to fix this? Or do I have to generate the annotations from scratch on my own using just simple entities?

I'm using the dimlines since the days of dxfwrite... so if there is no easy fix for this, it will mean quite some refactoring for me.

Thanks for any suggestions.

**Comment by mbway** [comment]
Do you have other software that is able to view dxf files that you could open your generated file in? I'm wondering whether the issue is in the generation of the dxf or the rendering of it.

Also if you could provide a small example showing the problem that would be useful

**Comment by k3rn3l3rr0r** [comment]
BricsCAD on Linux looks OK:
![Snímek z 2021-01-20 17-34-20](https://user-images.githubusercontent.com/3663357/105198119-fd91e480-5b45-11eb-8b63-fe79ba0fdd8a.png)

VariCAD on Linux looks OK:
![Snímek z 2021-01-20 17-31-00](https://user-images.githubusercontent.com/3663357/105198208-18fcef80-5b46-11eb-90b5-26b26969f375.png)

QCAD on Linux looks OK:
![Snímek z 2021-01-20 17-40-27](https://user-images.githubusercontent.com/3663357/105198835-c40da900-5b46-11eb-834e-cc14f57bbee8.png)


QCAD on Mac looks OK:
![2021-01-20_](https://user-images.githubusercontent.com/3663357/105198598-798c2c80-5b46-11eb-822c-8bc2961c4d30.png)


FreeCAD on Linux does not render at all:
![Snímek z 2021-01-20 17-41-00](https://user-images.githubusercontent.com/3663357/105199042-fae3bf00-5b46-11eb-88f3-31f928a7ee5b.png)



PDF output is offset to the right:
![Snímek z 2021-01-20 17-44-28](https://user-images.githubusercontent.com/3663357/105201206-45fed180-5b49-11eb-8c6c-30090d1dd95e.png)


I'm sorry that the example files are different in most cases, but the basic idea stays the same.

**Comment by mbway** [comment]
could this be related to https://github.com/mozman/ezdxf/discussions/310 ? have you tried the latest code from git?

wait, never mind. You were worried about the text and not the arrows overhanging, although it does look like they are in the image you posted (which should be fixed now)

if you post the dxf file that you're showing above I can take a look at how it decides where to put the text

**Comment by k3rn3l3rr0r** [comment]
Thanks a lot!
It's not possible to upload it here, so I will point you to a gist I've created for this purpose: https://gist.github.com/k3rn3l3rr0r/e17e5cfb08f9bd6f43725b2a3f485ed3

This file was generated in version 14.2, but I checked with 15.1 too and the PDF result is the same.
I will try it with git master codebase presumably tomorrow. Will report on #310 status too.

**Comment by mbway** [comment]
this is very strange. When I wrote the text layout code I created a test file to contain all the scenarios I could think of and I made sure this file rendered correctly (with respect to AutoCAD)
here is the file rendered. pretty much everything except vertical text is positioned correctly
![text](https://user-images.githubusercontent.com/4923501/105255445-77909080-5b7b-11eb-96f9-6a8fd4a7ccda.png)
The relevant part to this issue is the grid of different `TEXT` entities with different alignments set:
![text](https://user-images.githubusercontent.com/4923501/105255540-adce1000-5b7b-11eb-9719-e7660b44310b.png)
as you can see, with the current logic, every possible alignment case is positioned correctly. The current logic is:
```python
def _get_alignment(text: AnyText) -> Alignment:
    if isinstance(text, Text):
        # for the purposes of rendering, the anchor is always the bottom left, text.get_align() should be ignored
        return HAlignment.LEFT, VAlignment.BASELINE
    ...
```
but this must not always be the case since your file does render with the text in the wrong place.
here is what ezdxf reports as the alignment for your text entities:
```
MIDDLE_CENTER "860"
MIDDLE_CENTER "820"
TOP_LEFT "heading"
BOTTOM_LEFT "test"
```
if the logic is changed to `MIDDLE_CENTER` -> `(HAlignment.CENTER, VAlignment.LOWER_CASE_CENTER)` then your file looks correct
![img](https://user-images.githubusercontent.com/4923501/105256099-f4703a00-5b7c-11eb-885f-968bb46b1717.png)

but the text stress test is now incorrect
![text](https://user-images.githubusercontent.com/4923501/105255950-a0fdec00-5b7c-11eb-97b0-c6b5b1c96cf4.png)

I will have to take a closer look to see if there are any differences in the text entities that can be used to distinguish these cases

**Reply by mozman** [reply]
Added support for text alignments `FIT` and `ALIGNED`:

![image](https://user-images.githubusercontent.com/4230966/105451152-48ae1380-5c7c-11eb-9f3b-a926fab9e378.png)

But I have no idea how to remove the gap between the text and the definition line:

![image](https://user-images.githubusercontent.com/4230966/105451225-73986780-5c7c-11eb-8760-458dc58feea0.png)

**Reply by mozman** [reply]
If you want to place the text on the base line you have to use the BASELINE alignment ;-)

![image](https://user-images.githubusercontent.com/4230966/105455824-70a17500-5c84-11eb-9b3e-61c294e147cb.png)

**Comment by mbway** [comment, marked as answer]
I think I figured it out. Looks like to handle this case the insert should be ignored and the align_point should be used instead. In most cases it looks like the insert is set so that it lies at the left-baseline of the text (probably the CAD application takes care of this normally but the dxf file is valid if this is not the case).
It looks like the CAD applications trust the alignment point more than the insert location for where to place the text, so ezdxf should do the same.

I determined this from looking at the dxfattributes:
```
stress test file:
{'handle': '4E6', 'owner': '1F', 'layer': '0', 'insert': Vec3(3674.063120083396, 942.1986936556514, 0.0), 'height': 25.0, 'text': 'Xpba text', 'style': 'Simplex_100', 'halign': 1, 'align_point': Vec3(3765.134548654824, 954.6986936556514, 0.0), 'valign': 2}


provided file:
{'handle': None, 'owner': None, 'layer': 'DIMENSIONS', 'color': 6, 'insert': Vec3(430.0, -92.5, 0.0), 'height': 75.0, 'text': '860', 'style': 'OpenSansCondensed-Light', 'halign': 1, 'align_point': Vec3(430.0, -92.5, 0.0), 'valign': 2, 'width': 1.0, 'rotation': 0.0, 'extrusion': Vec3(0.0, 0.0, 1.0)}
```

I've opened a pull request with my fix: https://github.com/mozman/ezdxf/pull/320

**Comment by k3rn3l3rr0r** [comment]
Well, from question to fix available in the repo within 20 hours... Talk about open source magic.

I tried on top of your branch text_alignment (0.15.2a0) and this fixes my issue. All dimension texts are now positioned properly in the PDF.
Regarding the #310, that seems to be solved too, the dimticks are now ending in the right spots.

I will wait how it will turn out with the MR #320. Many thanks for your work!
